### DISCUSSION TITLE
ANTLR memory allocations look excessive?

### DISCUSSION THREAD

**Discussion by SimonSntPeter** [antlr/antlr4#3765] [General]
For a tiny little parsing input in C# I noticed it took a full second to start up. Using the profiler I looked at the memory allocated and found that it seems to be doing an gigantic amount of object generation, which then has to be cleaned up by the GC.

Here is my input, 413 bytes of it:

```
{
    LDB_run_internal_self_tests := false;
--    LDB_run_internal_self_tests := true;
};

create table not null orders
(
    onum int pkey,
    oname varchar(30),
    cnum int
);

select onum, count(oname) as nn
from orders
where onum = 23
group by onum
having count(oname) = 999
;

select onum, count(oname) as nn
from orders
where onum = 123
group by onum
having count(oname) = 888
;

```

Attached is an image of the profiler's output. The top graph I guess is the allocation nursery, and you can see what appeared to be five garbage collections in just that. I don't know what the bottom graph is but it's going to 48 meg.

Below that are the number of allocations from antlr, I have no idea how it's possible it could be doing this much.

A while back I did some profiling with the Python output, and tentatively concluded that the slowdown was something to do with the number of objects allocated, which makes sense because Python uses a skanky expensive reference counting GC and a fallback mark-sweep if necessary for cycles, whereas Java and C# seem to have much more efficient generational collectors.

I don't have time to dig into this, but I would like other people's opinion. Can anyone reproduce this?

Please note that the stuff I'm parsing actually produces an AST inside and a few other bits and pieces, but if you look at the breakdown in the attached image, you'll see that it really is utterly overwhelmed by antler-related objects.

Keep in mind I may be doing something wrong, but given that, any thoughts?

ja
![Capture_annot](https://user-images.githubusercontent.com/16326873/175988201-5e99a1ee-4a2c-4417-ba18-dc26f9b65b31.JPG)
n

**Comment by KvanTTT** [comment]
I am afraid this data can not be shrunk a lot, because ATN is a static representation of grammar. It's encoded into generated code, it's deserialized on loading, and it's used during lexing/parsing.

But it can be shrunk moderately if use ATN optimizations, for instance, this one: https://github.com/antlr/antlr4/pull/3676

**Comment by SimonSntPeter** [comment]
Hi, thanks for this.
It does leave me with a number of questions below:

> Because ATN is a static representation of grammar

I thought antler was recursive descent, and that C# output was really the encoding of the grammar? (I should know more about ALL*, I don't).

Putting that aside, that explains part of what I saw, which is the one part (ATN transition it might have been) didn't really grow when you had more input to parse, but other stuff did. For example, the very topmost item, ATN config drastically increased. I'll have to do some more tests tomorrow and post evidence, not tonight.

But it also fails to explain other things for example ATN.Transation enumerator being allocated in huge numbers, and nearly 20,000 sharpen.bitset allocations? That just makes no sense to me.

So, a question, is it possible to reuse a parser once constructed without throwing it away and recreating it from scratch? For example for me the code looks like this:


```
public static (LDBParser, ErrorListener<IToken>) makeParser(string inputStr) {
    var str = new AntlrInputStream(inputStr);
    var lexer = new LDBLexer(str);
    var tokens = new CommonTokenStream(lexer);
    var parser = new LDBParser(tokens);
    var errListener = new ErrorListener<IToken>(parser, lexer, tokens);
    parser.AddErrorListener(errListener);
    parser.AddErrorListener(new DiagnosticErrorListener(false));
    lexer.AddErrorListener(new ErrorListener<int>(parser, lexer, tokens));
    parser.Interpreter.PredictionMode = PredictionMode.LL_EXACT_AMBIG_DETECTION;

    return (parser, errListener);
}
```

and I do that every time.  If it's so expensive than I'd rather recycle than re-create from scratch - a one second startup if you're doing huge numbers of these will be crippling.

I'll look into this tomorrow and try and give further breakdown.

jan

**Reply by KvanTTT** [reply]
> Putting that aside, that explains part of what I saw, which is the one part (ATN transition it might have been) didn't really grow when you had more input to parse, but other stuff did. For example, the very topmost item, ATN config drastically increased. I'll have to do some more tests tomorrow and post evidence, not tonight.

Yes, another part is dynamic and it's being allocated more and more during program running. Responsible classes are located here: https://github.com/antlr/antlr4/tree/dev/runtime/Java/src/org/antlr/v4/runtime/dfa
 You can reset the cache by the way described on StackOverflow: https://stackoverflow.com/a/28843647/1046374.

> So, a question, is it possible to reuse a parser once constructed without throwing it away and recreating it from scratch? For example for me the code looks like this:

Dynamic info should be cached anyway. I don't know why it's being recreated on every construction. Maybe because of `LL_EXACT_AMBIG_DETECTION` mode, I've never used it.

**Comment by kaby76** [comment]
ATNConfig() constructors are not called during deserialization. But I doubt deserialization is at all a performance problem. For plsql--the largest grammar we have with many thousands of rules--deserialization takes ~0.05s for each atn, lexer and parser (Ryzen 7 2700 standard clock speed, 16GB DDR4, SSD), out of ~2s (plsql/examples/aggregate01.sql). I would first eliminate grammar issues. Check the ambiguity with Intellij on the small input. For example, for plsql, there's a max k of 26!! associated with unary_expression, and lots of ambiguity in atom. Nothing good can happen with numbers like these. That's why it takes 2s to parse a 73-line example. (The Intellij IDE also had a fatal error, but that's another issue.)

Remember: Antlr gives one a lot of rope to hang yourself. Everything starts with a quality grammar.

**Comment by KvanTTT** [comment]
BTW, the similar issue is existed: https://github.com/antlr/antlr4/issues/821

**Reply by parrt** [reply]
There is also the discussion of freeze drying the ATN state so we can't be reloaded, to avoid the start up cost. Don't have a link handy.

**Comment by SimonSntPeter** [comment]
Excuse delay in responding, just had my first covid.

Will skip gathering data and do as Ken Domino said and check lookahead. Also thanks to Ivan Kochurkin, will check out links. Will report back if I find anything noteworthy.

**Comment by SimonSntPeter** [comment]
I recall this being mentioned.

I think Ken Domino's point is likely the answer, that I've just written 
myself a very suboptimal grammar and should look at that first.

jan

**Reply by kaby76** [reply]
BTW, if you have an aversion to IDEs like I do, I've added a ["trperf" tool to Trash](https://github.com/kaby76/Domemtech.Trash/tree/main/trperf#readme) that outputs the perf table in exactly the same format as you would see in the Intellij Antlr plugin. And, it works with grammars that have semantic predicates. My plan (long term) is to use that to drive a tool to optimize the grammar.
```
$ # Compute perf stats, get rid of title header line, sort on "max-k", and output as a table.
$ trperf examples/aggregate01.sql | tail -n +2 | sort -k6 -n -r | column -t | head
Time to parse: 00:00:01.3121274
1442  unary_expression                              107  495696   85   26  5   5   0
1258  table_ref                                     21   392604   63   11  17  5   0
1261  table_ref_aux                                 21   398228   38   9   20  0   0
1458  atom                                          102  1387376  361  5   80  80  0
1612  tableview_name                                22   592153   42   4   16  5   0
1461  string_function                               3    32510    12   4   3   3   0
1582  table_alias                                   19   334882   57   3   19  19  0
1581  column_alias                                  29   512669   24   3   8   0   0
1580  column_alias                                  29   567477   76   3   29  29  0
1462  string_function                               3    28058    9    3   3   3   0
```

**Reply by KvanTTT** [reply]
> My plan (long term) is to use that to drive a tool to optimize the grammar.

What do you mean by optimizing the grammar? Can such a functionality be integrated into ANTLR tool?

**Reply by kaby76** [reply]
Not sure, just some ideas to explore. We know of two [refactorings that can help](https://tomassetti.me/improving-the-performance-of-an-antlr-parser):

* Grouping might improve performance because multiple calls to AdaptivePredict() will be unified ("Consolidating sub rules").
* Expressions where operator precedence is implemented as a change of rules can be rewritten in Antlr as an ambiguous rule with alts ("Handling expressions").

Since refactoring across a grammar is combinatorial, it likely has to be guided by performance.

**Reply by KvanTTT** [reply]
Actually, I'd consolidate our efforts since it looks like it can be implemented inside ANTLR tool.

**Reply by parrt** [reply]
Just remember that any change to the ATN or algorithm have huge inertia  with me;)

**Reply by KvanTTT** [reply]
BTW, some issues should be considered during consolidation. For instance, [we've had](https://github.com/antlr/grammars-v4/pull/2608) a case where `?` operator cannot be applied since it affects parsing behavior:

```c
structDeclaration
    :   specifierQualifierList structDeclaratorList ';'
    :   specifierQualifierList ';'
    |   staticAssertDeclaration
    ;
```

**Reply by kaby76** [reply]
```
structDeclaration // The first two rules have priority order and cannot be simplified to one expression.
    :   specifierQualifierList structDeclaratorList ';'
    |   specifierQualifierList ';'
    |   staticAssertDeclaration
    ;
```
vs
```
structDeclaration
    :   specifierQualifierList structDeclaratorList? ';'
    |   staticAssertDeclaration
    ;
```
Input `struct A { int x;};`

We need to be careful in refactoring with ambiguous parses. This string has two different derivations. Trash can do this refactoring, which is a grouping between the first and second alts. It's implemented using an N-way diff among all or some of the alts of a rule.

But, in this specific example, this is what happens when one does not follow the [Spec](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf), `struct-declaration : specifier-qualifier-list struct-declarator-list ;`. [It was in the initial version of the grammar](https://github.com/antlr/grammars-v4/blob/15df18159653f7ae2963cc73bec7d2b8939a9363/c/C.g4#L256).

**Reply by parrt** [reply]
also note that ambiguity is undecidable statically.

**Reply by kaby76** [reply]
> also note that ambiguity is undecidable statically.

Of course. But, given a finite set of inputs, ambiguity is decidable for the grammar under that set of strings. I think we can assume that we have a large test suite that would exercise the ambiguity.

**Reply by KvanTTT** [reply]
I don't completely understand how `?` operator works. Intuitively, `b?` should be treated either as `(b | )` or `( | b)`. Consider the following grammar:

```g4
test1
    : b? c
    ;

test31
    : (b | ) c
    ;

test32
    : ( | b) c
    ;

b: ID;
c: ID+;
ID: [a-z]+;
WS: []+ -> skip;
```

Rules `test` and `test31` return the following tree:

![parseTree1](https://user-images.githubusercontent.com/1150330/178119402-79d4cf97-c917-4bdf-96f4-a7df902d1677.png)

`test32` returns the following:

![parseTree2](https://user-images.githubusercontent.com/1150330/178119406-a5792b28-e32e-4fb6-a13d-8ab975d2eebf.png)

Ok, let `b?` is treated as `(b | )`. In this case, the rule

```g4
structDeclaration
    :   specifierQualifierList structDeclaratorList? ';'
    |   staticAssertDeclaration
    ;
```

should be equivalent to

```g4
structDeclaration
    :   specifierQualifierList structDeclaratorList ';'
    :   specifierQualifierList ';'
    |   staticAssertDeclaration
    ;
```

And deoptimization is not required. But it's not. It looks like `?` operator works inconsistently.

**Comment by NagyGa1** [comment]
We see a similar pattern on the c++ runtime. Here is a recent profile:

![image](https://user-images.githubusercontent.com/7298930/227062419-af1aacc3-6c75-4c18-8612-417e259f84fe.png)

It is ~8500 total samples, out of which (613+530+870+615+356)=2984 is memory allocation / deallocation / dynamic cast.

Probably not easy to move the existing code to use the stack instead of the heap, or some fixed structure, but would give a big boost.

The dynamic cast normally can be replaced with virtual methods (which still have one virtual table lookup in the background compared to non-virtuals).

These were small trees with 1-3 leaves.

**Reply by jimidle** [reply]
What grammar are you using?

**Reply by NagyGa1** [reply]
Sorry, I was wrong, here is the grammar, probably far from optimal. We probably tried to use the wrong tool for the job as well.

```antlr
grammar NumberValue;

start: (
		date
		| time_value
		| date_time
		| float_
		| int_
		| currency
		| exp
		| percentage
	) EOF;

date_time: date time_value;

date: (
		(INT ('-' | '/') INT ('-' | '/') INT)
		| (INT ('-' | '/') INT)
		| (INT STRING INT)
		| (STRING INT)
		| (INT STRING)
		| (INT '-' STRING '-' INT)
	);

time_value: ( (INT ':' INT ':' (FLOAT | INT)) | (INT ':' INT));

float_: (FLOAT | LPAREN FLOAT RPAREN | '-' FLOAT);

int_: ('-'? INT | LPAREN INT RPAREN); //('-'? INT | LPAREN INT RPAREN);

currency: CURRENCY (INT | FLOAT) (COMMA INT)* (PERIOD INT+)?;

exp: ('-'? (INT | FLOAT) EXP ('-'? INT));

percentage: P_CENT* ('-'? INT |'-'? FLOAT) P_CENT+;


EXP: 'e' | 'E';
CURRENCY: '$' | '€'; //rem '¥'
P_CENT: '%';
COMMA: ',';
PERIOD: '.';
COLON: ':';
LPAREN: '(';
RPAREN: ')';
INT: [0-9]+;
STRING: [A-Za-z]+;
FLOAT: INT PERIOD INT+;
WS: [ \t\r\n]+ -> skip;
```

**Reply by kaby76** [reply]
Why is any of this written in parser rules? It should all be in the lexer.

**Reply by NagyGa1** [reply]
Yes, noted. The feedback I'm trying to make here is that the current c++-runtime could be a lot faster without new SomeClass() and the likes.

**Comment by jimidle** [comment]
Well unless you’re parsing sone large input, then all measurements will be
dominated by initialization.

**Reply by NagyGa1** [reply]
Yes, noticed that all results seem to have a 2000ns baseline.

**Comment by jimidle** [comment]
Are you reusing the parser and lexer?

**Reply by NagyGa1** [reply]
No. Should? (Trying now, will report back)

**Reply by jimidle** [reply]
Also, you are right that if this all you are parsing, then there are better tools for this.

But your grammar is ambiguous as well. Refactoring will all you to use SLL mode, which will be more efficient, possibly by a lot. Hmm - maybe that grammar can use SLL mode. You can try it and see.

**Reply by jimidle** [reply]
For just dates, then this might better for you: https://github.com/HowardHinnant/date

**Reply by NagyGa1** [reply]
Thank you. Couldn't figure how to switch the c++-runtime into SLL mode, if you could send me to the right direction?

**Reply by jimidle** [reply]
I think this is what you need:

```c++
myparser = new MyParser(<tokensource>)
myparser->getInterpreteratn::ParserATNSimulator()
->setPredictionMode(atn::PredictionMode::SLL);
```

After that first new, you can just set a new token stream on it. This will mean you don't deserialize the ATN every time you parse

**Reply by NagyGa1** [reply]
Thank you Sir, will report back.

**Reply by NagyGa1** [reply]
The overhead came down from `~2000ns` to `~1500ns`, which is important for us even if we use Antlr4 only on larger grammars / more complexity. SLL did not produce an eyeball-visible difference in this case.

Thank you, was very helpful.

**Reply by NagyGa1** [reply]
...actually more like from `~1500ns` to `~1000ns`.

**Comment by jimidle** [comment]
You’re welcome

**Comment by dhaumann** [comment]
For the C++ runtime, I juts opened issue #4518, maybe that's interesting to some readers here.
